src/ripple/app/paths/impl/BookStep.cpp

```diff
@@ -38,14 +38,144 @@
+    , selfCrossingQualityLimit_ (getQuality (limitQuality))
+    {
+    }
+
+    bool limitSelfCrossQuality (AccountID const& strandSrc,
+        AccountID const& strandDst, TOffer<TIn, TOut> const& offer,
+        boost::optional<Quality>& ofrQ, FlowOfferStream<TIn, TOut>& offers,
+        bool const offerAttempted) const
+    {
+        // Special rule for offer crossing.  If:
+        //   a. We're offer crossing using default path (no autobridging), and
+        //   b. The offer's quality is at least as good as our quality, and
+        //   c. We're about to cross one of our own offers, then
+        //   d. Delete the old offer.
+        // See explanation in the testSelfCrossOffer1() unit test.
+        if (defaultPath_ && offer.quality() >= selfCrossingQualityLimit_ &&
+            strandSrc == offer.owner() && strandDst == offer.owner())
+        {
+            // Remove this offer even if no crossing occurs.
+            offers.permRmOffer (offer.key());
+
+            // If no offers have been attempted yet then it's okay to move to
+            // a different quality.
+            if (!offerAttempted)
+                ofrQ = boost::none;
+
+            // Return true so the current offer will be deleted.
+            return true;
+        }
+        return false;
+    }
+
+    // For offer crossing don't pay the transfer fee if alice is paying alice.
+    // The justification here is that the funds are not really changing hands;
+    // they go from alice to alice.  So it is inappropriate to charge a
+    // transfer fee.  Payments are too complicated to apply this rule uniformly.
+    std::uint32_t getOfrInRate (Step const* prevStep,
+        TOffer<TIn, TOut> const& offer, std::uint32_t trIn) const
+    {
+        auto const srcAcct = prevStep ?
+            prevStep->directStepSrcAcct() :
+            boost::none;
+
+        return                                        // If offer crossing
+            srcAcct &&                                // && prevStep is DirectI
+            offer.owner() == *srcAcct                 // && src is offer owner
+            ? QUALITY_ONE : trIn;                     // then rate = QUALITY_ONE
+    }
+
+    // See comment on getOfrInRate().
+    std::uint32_t getOfrOutRate (
+        Step const* prevStep, TOffer<TIn, TOut> const& offer,
+        AccountID const& strandDst, std::uint32_t trOut) const
+    {
+        return                                        // If offer crossing
+            prevStep && prevStep->bookStepBook() &&   // && prevStep is BookStep
+            offer.owner() == strandDst                // && dest is offer owner
+            ? QUALITY_ONE : trOut;                    // then rate = QUALITY_ONE
+    }
+
+private:
+    bool const defaultPath_;
+    Quality const selfCrossingQualityLimit_;
+};
+
+//------------------------------------------------------------------------------
+
+template<class TIn, class TOut, template <class In, class Out> class Policy>
```

Comment:
My personal preference is to leave the `Policy` as a `template template`.  The `Policy` actually is a `template`.  Also, the `template template` implementation guarantees that the `TIn` and `TOut` parameters of the `Policy` and the `BookStep` are the same, which is exactly what we want.  Yes, since we're doing explicit instantiation we could just do the right thing at the construction site.  But I think this approach is slightly more maintainable and more representative of what's actually up.

That said, I'm willing to go with your approach if you feel strongly.  Since `flow` is primarily your baby, you need to be comfortable with what's here.